druid/druid.sh

```diff
@@ -0,0 +1,526 @@
+# distributed under the License is distributed on an "AS-IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# This script installs Druid on a Google Cloud
+# Dataproc cluster.
+#
+# To use this script, you will need to configure the following variables to
+# match your cluster. For information about which software components
+# (and their version) are included in Cloud Dataproc clusters, see the
+# Cloud Dataproc Image Version information:
+# https://cloud.google.com/dataproc/concepts/dataproc-versions
+
+set -euxo pipefail
+
+export DRUID_VERSION='0.13.0-incubating'
+export DRUID_HOME="/opt/druid"
+export DRUID_DIR="${DRUID_HOME}/apache-druid-${DRUID_VERSION}"
+
+function err() {
+  echo "[$(date +'%Y-%m-%dT%H:%M:%S%z')]: $@" >&2
+  return 1
+}
+
+function make_druid_home() {
+  mkdir -p ${DRUID_HOME}
+}
+
+function download_druid() {
+  cd ${DRUID_HOME}
+  wget -nc http://ftp.man.poznan.pl/apache/incubator/druid/${DRUID_VERSION}/apache-druid-${DRUID_VERSION}-bin.tar.gz
+  tar -xzf apache-druid-${DRUID_VERSION}-bin.tar.gz
+}
+
+function extract_runtime_property(){
+  local file="${1}"
+  local parameter="${2}"
+  echo $(cat ${file} | grep ${parameter} | grep -o -E '[0-9]+')
+}
+
+function calculate_xmx(){
+  local processing_buffer=${1}
+  local num_threads=${2}
+  local num_merge_buffers=${3}
+
+  echo $(( ${processing_buffer} * (${num_merge_buffers} + ${num_threads}) + 1 ))
+}
+
+function prepare_environment_services(){
+  local metastore_instance
+  local password_argument
+  metastore_instance="$(/usr/share/google/get_metadata_value attributes/hive-metastore-instance)"
+  if [ -z "${metastore_instance}" ];then
+    systemctl start mysql
+    password_argument='-proot-password'
+  else
+    password_argument=''
+  fi
+  mysql -u root ${password_argument} -e "CREATE DATABASE druid DEFAULT CHARACTER SET utf8;" \
+    || echo "Database already exists"
+  mysql -u root ${password_argument} -e "GRANT ALL ON druid.* TO 'druid' IDENTIFIED BY 'diurd';" \
+    || echo "Cannot grant priviliges to druid user"
+}
+
+function max() {
+   [ "$1" -gt "$2" ] && echo $1 || echo $2
+}
+
+function modify_jvm_config(){
+  local service_name="${1}"
+  local cores
+  local processing_buffer
+  local num_threads
+  local num_merge_buff
+  local xmx
+  local xx
+  processing_buffer=$(extract_runtime_property \
+    ${DRUID_DIR}/conf/druid/${service_name}/runtime.properties druid.processing.buffer.sizeBytes)
+  if [[ ${processing_buffer} == "" ]];then
+    # Default value 1GB
+    processing_buffer=1073741824
+  fi
+  cores=$(grep -c ^processor /proc/cpuinfo)
+  cores=$((${cores}-1))
+  num_threads=$(max 1 ${cores})
+  if [[ ${num_threads} < 4 ]];then
+    num_merge_buff=2
+  else
+    num_merge_buff=$(max 2 ${num_threads}/4)
+  fi
+  xmx=$(calculate_xmx ${processing_buffer} ${num_threads} ${num_merge_buff})
```

Comment:
Memory should be computed based on the size of nodes.

Reply:
@aniket486 Let me know what do you think about this simple change I've applied with last commit. I observed that when RAM size is lower than 7 GB for node then broker cannot be launched because lack of memory.